packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * Appending the wp-block-heading to before rendering the stored `core/heading` block contents.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading block content.
+ *
+ * For example, the following block content:
+ *  <h2 class="align-left">Hello World</h2>
+ *
+ * Would be transformed to:
+ *  <h2 class="align-left wp-block-heading">Hello World</h2>
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+	// $pattern matches the opening tag of the heading element .
+	$pattern = '/
+        ^(\s*)                               # Any leading whitespace.
+        <(?<tag_name>
+            h[1-6]                           # The opening tag...
+            (?=\s|>)                         # ...followed by a whitespace or >
+                                             # ?= means a "lookahead"
+        )
+        (?<before_class>                     # Any attributes prior to "class"
+            [^>]*?                           # Non-greedy match of any character except ">"
+        )
+        (?:\s+                               # Until we find the class attribute, if any
+            class=(?P<quote>[\'"])           # The quote character
+            (?<class_name>.*?)               # Non-greedy match of any character
+            (\k{quote})                      # Until we find that quote character again
+            (?<after_class>[^>]*?)           # The rest of the tag
+        )?
+        >                                    # The closing tag
+    /xm';
+
+	return preg_replace_callback(
+		$pattern,
+		function ( $matches ) {
+			// Parse the existing class names.
+			$current_class_attr = ! empty( $matches['class_name'] ) ? $matches['class_name'] : '';
+
+			// If wp-block-heading is already included, there's no need to add it again.
+			$class_to_add = 'wp-block-heading';
+			if ( preg_match( "/\b$class_to_add\b/", $current_class_attr ) ) {
+				return $matches[0];
+			}
```

Comment:
I think the `\b` is not a blocker here, but just so we're aware, it captures probably more than we want, including things like punctuation.

also fun fact! PHP has supported list restructuring for a loooong time

```php
list( '0' => $fullmatch, 'quote' => $quote, 'class_name' => $class_name ) = $matches;
```

I like using this as an aid to comprehension though if a capture group is optional you have to either tell the PCRE engine to include it anyway _or_ grab that capture group like normal outside of `list()` where we can `if ( isset( ) )` it